src/lib/uuid/tt_uuid.h

```diff
@@ -182,6 +184,28 @@ tt_uuid_str(const struct tt_uuid *uu);
 int
 tt_uuid_from_strl(const char *in, size_t len, struct tt_uuid *uu);
 
+/**
+ * Error payload couldn't be implemented in libcore because requires UUID type
+ * and its methods. That would be a cyclic dep. Thus in places where need to use
```

Comment:
I don't like dependency of uuid on error_payload either. Uuid library has already provided msgpack serialization, why it must do anything else?

What about adding `error_payload_get_X` etc as a macro that takes structure name? all out mp serializers should have similar API (as I see only `mp_sizeof_uuid()` somehow does not take `struct tt_uuid *` as an argument). It would allow simply add more structures (datetime, who knows?).

Perhaps a macro that defines a static functions for that would be better.

Reply:
> I don't like dependency of uuid on error_payload either. Uuid library has already provided msgpack serialization, why it must do anything else?

I don't like it either but didn't find a better solution without introduction of cyclic dependencies and without merging libuuid into libcore. Now it is technically correct, because neither of these drawbacks are present. libuuid just uses libcore and extends its API a bit. For instance, box does the same with `error.h/.cc` - adds more errors and their methods on top of libcore.

There is another alternative though - move all extensions to box `error.h/.cc`. The drawback will be that you won't be able to use `error_set/get_uuid/datetime/decimal/...` in libraries like raft, swim. Don't like that TBH.

> What about adding error_payload_get_X etc as a macro that takes structure name?

Do you mean this?
```C
error_set(tt_uuid, value);
error_set(int, value);
error_set(decimal, value);
...
```
@sergepetrenko @locker what do you think? We could do that too. As for `mp_sizeof_uuid()` not taking any arguments - we could fix that. It is not public API anyway.

Reply:
> ```c
> error_set(tt_uuid, value);
> error_set(int, value);
> error_set(decimal, value);
> ...
> ```

I like this idea

Reply:
> ```c
> error_set(tt_uuid, value);
> error_set(int, value);
> error_set(decimal, value);
> ...
> ```

I don't understand how this would work with int/double/string:
- For string we need to store 0 at the end of the encoded msgpack data. Besides `mp_sizeof_str` takes the string length, not the string itself.
- For integers we have to encode positive integers as `MP_UINT` and negative integers as `MP_INT` plus handle possible overflow.
- For double we have to handle both `MP_FLOAT` and `MP_DOUBLE`.

Also, adding an unused argument to `mp_sizeof_uuid()` looks ugly.

Also, macros/templates would make the code more difficult to follow and modify IMO - we couldn't have any special cases/hacks in `error_payload_{get,set}_*` (like what you do for str).

If the issue is the dependency between uuid and error_payload (which I don't think is a problem BTW), we could solve it by adding one more compilation unit - `uuid_error_payload.[hc]`, but I wouldn't bother - I'd just moved `error_payload_{set,get}_uuid` to `error_payload.c` - I think it'd be totally fine for the error module to depend on uuid or decimal, as long as `error.h` doesn't need to include `uuid.h` and `decimal.h`.

Reply:
Yes, I thought exactly about `error_set(int, value);`. But I see that str, int, double, bool and tt_uuid have quite different encoders/decoders. Perhaps our msgpack library won't allow to avoid dependency.

@locker 
BTW even `mp_sizeof_bool` takes (and ignores) bool as an argument. This looks strange when look at one particular function but becomes robust when you look at entire library.

`uuid_error_payload.[hc]` looks better for me, but not perfect. I bother that having N types (int, float, uuid, decimal etc) that all has MP encoders/decoders and M consumers (error, tuple IDK) that requires MP serialization we need to write M*N adapters for each pair. That looks like dependency mess. But I don't see a good way out.

Perhaps we could make macros only for extension types?.
```c++
error_set(struct tt_uuid, value);
error_set(struct tt_decimal, value);
error_set(struct tt_datetime, value);
```

Reply:
> BTW even `mp_sizeof_bool` takes (and ignores) bool as an argument. This looks strange when look at one particular function but becomes robust when you look at entire library.
> 

It is weird. I'd drop that argument. We can't have unified `mp_sizeof/encode` functions anyway because of str and ext - they take the length of the object, not the object itself, in contrast to the rest of the functions.

> `uuid_error_payload.[hc]` looks better for me, but not perfect. I bother that having N types (int, float, uuid, decimal etc) that all has MP encoders/decoders and M consumers (error, tuple IDK) that requires MP serialization we need to write M*N adapters for each pair. That looks like dependency mess. But I don't see a good way out.
> 

The proliferation of headers doesn't look good, you're right. Let's put these functions in `error_payload.c`. As I said before, I think it's okay for the `error` module to depend on `decimal`, 'datetime', `uuid` - it wouldn't even create a circular dependency, because the latter don't set errors.

> Perhaps we could make macros only for extension types?.
> 
> ```c++
> error_set(struct tt_uuid, value);
> error_set(struct tt_decimal, value);
> error_set(struct tt_datetime, value);
> ```

I don't think that passing the struct keyword would work - how would you tear it off to use the struct name for `mp_sizeof/encode`? And `decimal` isn't a struct, unfortunately (though it should be fixed to allow forward declarations). I suggest not to spend time on this now, when we have just a couple of extensions. Note, we have exactly the same problem with `mpstream`, and nobody cares.

Reply:
Ok, it seems that macros won't work. I think that shows that our msgpack library cannot be abstracted.
IMO every dependency should be avoided. `uuid_error_payload.[hc]` would make `error_payload` and `tt_uuid` cross-independent, and `uuid_error_payload.[hc]` would be dependent on both. Why can't we do so?

Reply:
> `uuid_error_payload.[hc]` would make `error_payload` and `tt_uuid` cross-independent, and `uuid_error_payload.[hc]` would be dependent on both. Why can't we do so?

We can, but I wouldn't, because users would have to include this new header in addition to `error.h` and `uuid.h`, which is inconvenient. And it looks like self-deceit anyway: by avoiding *one* dependency (error -> uuid), we'd add *two* (uuid_error_payload -> uuid, uuid_error_payload -> error). I just noticed that uuid actually *does* depend on core - this is weird, I think we should either get rid of this dependency or, preferably, move uuid to core (decimal and datetime is a part of core BTW).

Reply:
I'd accept any variant.

Reply:
Done. I merged libuuid into libcore.